File: src/postprocess.cpp
 #include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

std::vector<int> nms(
	const std::vector<std::vector<float>> &boxes,
	const std::vector<float> &scores,
	float iou_thres = 0.45)
{
	int N = boxes.size();
	std::vector<float> x1(N), y1(N), x2(N), y2(N);
	for (int i = 0; i < N; i++)
	{
		x1[i] = boxes[i][0];
		y1[i] = boxes[i][1];
		x2[i] = boxes[i][2];
		y2[i] = boxes[i][3];
	}
	std::vector<float> areas(N);
	for (int i = 0; i < N; i++)
	{
		areas[i] = (x2[i] - x1[i]) * (y2[i] - y1[i]);
	}
	std::vector<int> order(N);
	for (int i = 0; i < N; i++)
	{
		order[i] = i;
	}
	std::sort(order.begin(), order.end(), [&scores](int a, int b)
			  { return scores[a] > scores[b]; });
	std::vector<int> keep;
	while (order.size() > 0)
	{
		int i = order[0];
		keep.push_back(i);
		if (order.size() == 1)
		{
			break;
		}
		std::vector<int> new_order;
		for (int k = 1; k < order.size(); k++)
		{
			int j = order[k];
      		float xx1 = std::max(x1[i], x1[j]);
      		float yy1 = std::max(y1[i], y1[j]);
      		float xx2 = std::min(x2[i], x2[j]);
     		float yy2 = std::min(y2[i], y2[j]);
			float w = max(0.0f, xx2 - xx1);
			float h = max(0.0f, yy2 - yy1);
			float inter = w * h;
			float iou_val = inter / (areas[i] + areas[j] - inter);
			if (iou_val <= iou_thres)
			{
				new_order.push_back(j);
			}
		}
		order = new_order;
		
	}
	return keep;
}

void postprocess(
	const float *output,
	std::vector<std::vector<float>> &boxes_out,
	std::vector<int> &class_out,
	std::vector<float> &score_out,
	float conf_thres = 0.25,
	float iou_thres = 0.45)
{
	const int NUM_ANCHORS = 8400;
	const int NUM_FEATURES = 84;
	vector<vector<float>> all_boxes(NUM_ANCHORS, vector<float>(4));
	vector<float> all_scores(NUM_ANCHORS);
	vector<int> all_classes(NUM_ANCHORS);
	for (int i = 0; i < NUM_ANCHORS; i++)
	{
		int offset = i * NUM_FEATURES;
		float cx = output[offset + 0];
		float cy = output[offset + 1];
		float bw = output[offset + 2];
		float bh = output[offset + 3];

		all_boxes[i][0] = cx - bw / 2.0f;
		all_boxes[i][1] = cy - bh / 2.0f;
		all_boxes[i][2] = cx + bw / 2.0f;
		all_boxes[i][3] = cy + bh / 2.0f;

		float max_score = 0.0f;
		int best_class = 0;
		for (int c = 0; c < 80; c++)
		{
			float score = output[offset + 4 + c];
			if (score > max_score)
			{
				max_score = score;
				best_class = c;
			}
		}
		all_scores[i] = max_score;
		all_classes[i] = best_class;
	}
	vector<vector<float>> filtered_boxes;
	vector<float> filtered_scores;
	vector<int> filtered_classes;
	for (int i = 0; i < NUM_ANCHORS; i++)
	{
		if (all_scores[i] > conf_thres)
		{
			filtered_boxes.push_back(all_boxes[i]);
			filtered_scores.push_back(all_scores[i]);
			filtered_classes.push_back(all_classes[i]);
		}
	}
	float img_w = 640.0;
	float img_h = 640.0;
	for (auto &box : filtered_boxes)
	{
		box[0] = std::clamp(box[0], 0.0f, img_w);
		box[1] = std::clamp(box[1], 0.0f, img_h);
		box[2] = std::clamp(box[2], 0.0f, img_w);
		box[3] = std::clamp(box[3], 0.0f, img_h);
	}
	vector<int> keep = nms(filtered_boxes, filtered_scores, iou_thres);

	for (auto v : keep)
	{
		boxes_out.push_back(filtered_boxes[v]);
		class_out.push_back(filtered_classes[v]);
		score_out.push_back(filtered_scores[v]);
	}
}

File: src/postprocess.py
"""
标准 YOLOv8 单输出后处理
======================
输入: NPU 输出 (1, 84, 8400)
  通道 0:  cx (像素坐标)
  通道 1:  cy (像素坐标)
  通道 2:  w (像素)
  通道 3:  h (像素)
  通道 4-83: 80 类别置信度 (已过 sigmoid)
输出: xyxy 像素坐标 + 类别 + 置信度

数据证明：box[6988] cx=449.0 cy=216.2 w=67.2 h=62.5
  → 这是像素坐标（640×640 letterbox 空间内），不是网格单位。
"""

import numpy as np


def nms(boxes, scores, iou_thres=0.45):
    """纯 numpy NMS"""
    x1, y1, x2, y2 = boxes[:, 0], boxes[:, 1], boxes[:, 2], boxes[:, 3]
    areas = (x2 - x1) * (y2 - y1)
    order = scores.argsort()[::-1]
    keep = []

    while order.size > 0:
        i = order[0]
        keep.append(i)
        if order.size == 1:
            break

        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])

        w = np.maximum(0.0, xx2 - xx1)
        h = np.maximum(0.0, yy2 - yy1)
        inter = w * h
        iou = inter / (areas[i] + areas[order[1:]] - inter)

        order = order[1:][iou <= iou_thres]

    return np.array(keep, dtype=np.int64)


def postprocess(output, conf_thres=0.25, iou_thres=0.45, img_size=(640, 640)):
    """
    输入: output = NPU 输出 (1, 84, 8400)
    返回: (boxes, classes, scores)
    """
    out = output[0].T  # (8400, 84)

    cx = out[:, 0]
    cy = out[:, 1]
    bw = out[:, 2]
    bh = out[:, 3]
    cls_raw = out[:, 4:]  # (8400, 80)

    # cxcywh → xyxy (已经是像素坐标，直接算)
    x1 = cx - bw / 2.0
    y1 = cy - bh / 2.0
    x2 = cx + bw / 2.0
    y2 = cy + bh / 2.0
    boxes = np.stack([x1, y1, x2, y2], axis=1)

    # 最高分类别
    scores = cls_raw.max(axis=1)
    class_ids = cls_raw.argmax(axis=1)

    # 置信度过滤
    mask = scores > conf_thres
    boxes = boxes[mask]
    scores = scores[mask]
    class_ids = class_ids[mask]

    if len(boxes) == 0:
        return np.array([]), np.array([]), np.array([])

    # 裁剪
    boxes[:, 0] = np.clip(boxes[:, 0], 0, img_size[0])
    boxes[:, 1] = np.clip(boxes[:, 1], 0, img_size[1])
    boxes[:, 2] = np.clip(boxes[:, 2], 0, img_size[0])
    boxes[:, 3] = np.clip(boxes[:, 3], 0, img_size[1])

    # NMS
    indices = nms(boxes, scores, iou_thres)
    return boxes[indices], class_ids[indices], scores[indices]
